/**
 * POP API
 *
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CodedConcept } from './coded-concept';
import { GenomicVariantAssessmentChoices } from './genomic-variant-assessment-choices';
import { GenomicVariantConfidenceChoices } from './genomic-variant-confidence-choices';
import { GenomicVariantClinicalRelevanceChoices } from './genomic-variant-clinical-relevance-choices';


export interface GenomicVariantCreate { 
    genomicHgvs?: string | null;
    dnaHgvs?: string | null;
    rnaHgvs?: string | null;
    proteinHgvs?: string | null;
    externalSource?: string | null;
    externalSourceId?: string | null;
    /**
     * Indicates the case of the patient who\' genomic variant is described
     */
    caseId: string;
    /**
     * Clinically-relevant date at which the genomic variant was detected and/or reported.
     */
    date: string;
    genePanel?: string | null;
    assessment?: GenomicVariantAssessmentChoices | null;
    confidence?: GenomicVariantConfidenceChoices | null;
    analysisMethod?: CodedConcept | null;
    clinicalRelevance?: GenomicVariantClinicalRelevanceChoices | null;
    genomeAssemblyVersion?: CodedConcept | null;
    molecularConsequence?: CodedConcept | null;
    copyNumber?: number | null;
    alleleFrequency?: number | null;
    alleleDepth?: number | null;
    zygosity?: CodedConcept | null;
    inheritance?: CodedConcept | null;
    coordinateSystem?: CodedConcept | null;
    clinvar?: string | null;
    /**
     * Gene(s) affected by this variant
     */
    genes?: Array<CodedConcept>;
}
export namespace GenomicVariantCreate {
}


